File: src/AIEmployeeSupport.Application/Services/AnalyticsService.cs
using AIEmployeeSupport.Application.DTOs.Analytics;
using AIEmployeeSupport.Application.DTOs.Common;
using AIEmployeeSupport.Application.Interfaces;
using AIEmployeeSupport.Application.Interfaces.Services;
using AIEmployeeSupport.Domain.Enums;

namespace AIEmployeeSupport.Application.Services;

public class AnalyticsService : IAnalyticsService
{
    private readonly ISupportQuestionRepository _questionRepository;
    private readonly IKnowledgeScenarioRepository _scenarioRepository;
    private readonly IKnowledgeCategoryRepository _categoryRepository;
    private readonly IAIRequestLogRepository _requestLogRepository;

    public AnalyticsService(
        ISupportQuestionRepository questionRepository,
        IKnowledgeScenarioRepository scenarioRepository,
        IKnowledgeCategoryRepository categoryRepository,
        IAIRequestLogRepository requestLogRepository)
    {
        _questionRepository = questionRepository;
        _scenarioRepository = scenarioRepository;
        _categoryRepository = categoryRepository;
        _requestLogRepository = requestLogRepository;
    }

    private static (DateTime? start, DateTime? end) NormalizeDateRange(DateTime? fromDate, DateTime? toDate)
    {
        DateTime? start = fromDate;
        DateTime? end = toDate;

        if (end.HasValue && end.Value.TimeOfDay == TimeSpan.Zero)
        {
            end = end.Value.Date.AddDays(1).AddTicks(-1);
        }

        return (start, end);
    }

    public async Task<OverviewAnalyticsDto> GetOverviewAsync(DateTime? fromDate = null, DateTime? toDate = null, CancellationToken cancellationToken = default)
    {
        var (start, end) = NormalizeDateRange(fromDate, toDate);
        var questionsTuple = await _questionRepository.GetAllAsync(1, int.MaxValue, null, start, end, cancellationToken);
        var questions = questionsTuple.Items.ToList();
        
        var today = DateTime.UtcNow.Date;
        
        var answered = questions.Count(q => q.Status == QuestionStatus.Answered && q.AnsweredByAI);
        var escalated = questions.Count(q => q.Escalated);
        var total = questions.Count;
        var processingTimes = questions.Where(q => q.ProcessingTimeMs.HasValue).Select(q => q.ProcessingTimeMs!.Value).ToList();
        var confidenceScores = questions.Where(q => q.ConfidenceScore.HasValue).Select(q => q.ConfidenceScore!.Value).ToList();

        return new OverviewAnalyticsDto
        {
            TotalQuestions = total,
            TodayQuestions = fromDate.HasValue || toDate.HasValue ? total : questions.Count(q => q.CreatedAt.Date == today),
            AnsweredCount = answered,
            NoAnswerCount = questions.Count(q => (!q.AnsweredByAI && q.Status != QuestionStatus.New) || q.Status == QuestionStatus.NoAnswer),
            EscalatedCount = escalated,
            AvgResponseTimeMs = processingTimes.Any() ? Math.Round(processingTimes.Average(), 2) : 0,
            AvgSimilarityScore = confidenceScores.Any() ? Math.Round(confidenceScores.Average(), 4) : 0,
            AnswerRate = total > 0 ? Math.Round((double)answered / total, 4) : 0
        };
    }

    public async Task<QuestionAnalyticsDto> GetQuestionAnalyticsAsync(DateTime? fromDate = null, DateTime? toDate = null, CancellationToken cancellationToken = default)
    {
        var (start, end) = NormalizeDateRange(fromDate, toDate);
        var questionsTuple = await _questionRepository.GetAllAsync(1, int.MaxValue, null, start, end, cancellationToken);
        var query = questionsTuple.Items.AsEnumerable();

        return new QuestionAnalyticsDto
        {
            Questions = query.Select(q => new QuestionAnalyticsItem
            {
                Id = q.Id,
                EmployeeId = q.EmployeeId,
                EmployeeName = q.Employee != null ? q.Employee.FullName : null,
                EmployeeEmail = q.Employee != null ? q.Employee.Email : null,
                QuestionText = q.QuestionText,
                Status = q.Status.ToString(),
                AnsweredByAI = q.AnsweredByAI,
                ConfidenceScore = q.ConfidenceScore,
                ProcessingTimeMs = q.ProcessingTimeMs,
                ScenarioName = q.Scenario != null ? q.Scenario.Name : null,
                CreatedAt = q.CreatedAt,
                CompletedAt = q.CompletedAt
            }).ToList()
        };
    }

    public async Task<PaginatedResponse<KnowledgeAnalyticsDto>> GetKnowledgeAnalyticsAsync(
        int page = 1, int pageSize = 10, DateTime? fromDate = null, DateTime? toDate = null, CancellationToken cancellationToken = default)
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 100);

        var (start, end) = NormalizeDateRange(fromDate, toDate);
        var (questionsTuple, _) = await _questionRepository.GetAllAsync(1, int.MaxValue, null, start, end, cancellationToken);
        var scenarioStats = questionsTuple
            .Where(q => q.ScenarioId.HasValue)
            .GroupBy(q => q.ScenarioId.Value)
            .ToDictionary(
                g => g.Key,
                g => new
                {
                    Count = g.Count(),
                    AvgSimilarityScore = g.Any(q => q.ConfidenceScore.HasValue)
                        ? g.Where(q => q.ConfidenceScore.HasValue).Average(q => q.ConfidenceScore ?? 0)
                        : 0,
                    LastUsed = g.OrderByDescending(q => q.CreatedAt).FirstOrDefault()?.CreatedAt
                });

        var (scenarios, totalCount) = await _scenarioRepository.GetAllAsync(page, pageSize, null, null, cancellationToken);

        var items = scenarios.Select(s =>
        {
            var hasStats = scenarioStats.TryGetValue(s.Id, out var stats) && stats != null;
            return new KnowledgeAnalyticsDto
            {
                ScenarioId = s.Id,
                ScenarioName = s.Name,
                CategoryName = s.Category?.Name ?? "غير مصنف",
                RetrievalCount = hasStats ? stats!.Count : 0,
                AvgSimilarityScore = hasStats ? Math.Round(stats!.AvgSimilarityScore, 4) : 0,
                LastUsed = hasStats ? stats!.LastUsed : null
            };
        }).ToList();

        return new PaginatedResponse<KnowledgeAnalyticsDto>
        {
            Items = items,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<PaginatedResponse<CategoryAnalyticsDto>> GetCategoryAnalyticsAsync(
        int page = 1, int pageSize = 10, DateTime? fromDate = null, DateTime? toDate = null, CancellationToken cancellationToken = default)
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 100);

        var categories = (await _categoryRepository.GetAllAsync(cancellationToken)).ToList();
        var (scenarios, _) = await _scenarioRepository.GetAllAsync(1, int.MaxValue, null, null, cancellationToken);

        var (start, end) = NormalizeDateRange(fromDate, toDate);
        var (questions, _) = await _questionRepository.GetAllAsync(1, int.MaxValue, null, start, end, cancellationToken);
        
        var scenariosList = scenarios.ToList();
        var questionsList = questions.ToList();
        var totalWithScenario = questionsList.Count(q => q.ScenarioId.HasValue);
        
        var allCategories = categories.Select(c =>
        {
            var catScenarioIds = scenariosList.Where(s => s.CategoryId == c.Id).Select(s => s.Id).ToHashSet();
            var count = questionsList.Count(q => q.ScenarioId.HasValue && catScenarioIds.Contains(q.ScenarioId.Value));
            return new CategoryAnalyticsDto
            {
                CategoryId = c.Id,
                CategoryName = c.Name,
                ScenarioCount = catScenarioIds.Count,
                QuestionCount = count,
                Percentage = totalWithScenario > 0 ? Math.Round((double)count / totalWithScenario * 100, 1) : 0
            };
        }).OrderByDescending(x => x.QuestionCount).ThenByDescending(x => x.ScenarioCount).ToList();

        var knownCategoryIds = categories.Select(c => c.Id).ToHashSet();
        var unassignedScenarios = scenariosList.Where(s => s.CategoryId == Guid.Empty || !knownCategoryIds.Contains(s.CategoryId)).Select(s => s.Id).ToHashSet();
        if (unassignedScenarios.Count > 0)
        {
            var unassignedCount = questionsList.Count(q => q.ScenarioId.HasValue && unassignedScenarios.Contains(q.ScenarioId.Value));
            allCategories.Add(new CategoryAnalyticsDto
            {
                CategoryId = Guid.Empty,
                CategoryName = "غير مصنف",
                ScenarioCount = unassignedScenarios.Count,
                QuestionCount = unassignedCount,
                Percentage = totalWithScenario > 0 ? Math.Round((double)unassignedCount / totalWithScenario * 100, 1) : 0
            });
        }

        var totalCount = allCategories.Count;
        var pagedItems = allCategories.Skip((page - 1) * pageSize).Take(pageSize).ToList();

        return new PaginatedResponse<CategoryAnalyticsDto>
        {
            Items = pagedItems,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<UnansweredAnalyticsDto> GetUnansweredAnalyticsAsync(
        int page = 1, int pageSize = 10, string sortOrder = "desc", DateTime? fromDate = null, DateTime? toDate = null, CancellationToken cancellationToken = default)
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 100);

        var (start, end) = NormalizeDateRange(fromDate, toDate);
        var unansweredQuestions = (await _questionRepository.GetUnansweredAsync(start, end, cancellationToken)).ToList();
        
        var query = unansweredQuestions
            .GroupBy(q => q.QuestionText.ToLowerInvariant().Trim())
            .Select(g => 
            {
                var latest = g.OrderByDescending(q => q.CreatedAt).First();
                return new UnansweredQuestionItem
                {
                    Id = latest.Id,
                    EmployeeId = latest.EmployeeId,
                    EmployeeName = latest.Employee != null ? latest.Employee.FullName : null,
                    EmployeeEmail = latest.Employee != null ? latest.Employee.Email : null,
                    QuestionText = latest.QuestionText,
                    Frequency = g.Count(),
                    FirstAsked = g.Min(q => q.CreatedAt),
                    LastAsked = g.Max(q => q.CreatedAt)
                };
            });

        query = string.Equals(sortOrder, "asc", StringComparison.OrdinalIgnoreCase)
            ? query.OrderBy(x => x.LastAsked)
            : query.OrderByDescending(x => x.LastAsked);

        var list = query.ToList();
        var totalCount = list.Count;
        var pagedItems = list.Skip((page - 1) * pageSize).Take(pageSize).ToList();

        return new UnansweredAnalyticsDto
        {
            Questions = pagedItems,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }
}

File: client/src/services/analyticsService.ts
import api from './api';

export interface AnalyticsOverview {
    totalQuestions: number;
    todayQuestions?: number;
    questionsToday?: number;
    answeredCount?: number;
    answeredQuestions?: number;
    noAnswerCount?: number;
    unansweredQuestions?: number;
    escalatedCount?: number;
    escalated?: number;
    avgResponseTimeMs?: number;
    avgResponseTime?: string;
    avgSimilarityScore?: number;
    answerRate?: number;
    successRate?: number;
}

export interface QuestionAnalytics {
    id?: string;
    date?: string;
    count?: number;
    questionText?: string;
    status?: string;
    answeredByAI?: boolean;
    confidenceScore?: number;
    processingTimeMs?: number;
    scenarioName?: string;
    createdAt?: string;
}

export interface KnowledgeAnalytics {
    id?: string | number;
    scenarioId?: string;
    categoryId?: string | number;
    categoryName?: string;
    scenarioName?: string;
    usageCount?: number;
    retrievalCount?: number;
    avgSimilarityScore?: number;
}

export interface CategoryAnalytics {
    categoryId: string;
    categoryName: string;
    scenarioCount: number;
    questionCount: number;
    percentage: number;
}

export interface PaginatedResult<T> {
    items: T[];
    totalCount: number;
    page: number;
    pageSize: number;
    totalPages: number;
}

export interface UnansweredQuestion {
    id: string | number;
    question?: string;
    questionText?: string;
    timestamp?: string;
    createdAt?: string;
    firstAsked?: string;
    frequency?: number;
    employeeId?: string;
    employeeName?: string;
    employeeEmail?: string;
}

const analyticsService = {
    getOverview: async (startDate?: string, endDate?: string): Promise<AnalyticsOverview> => {
        const response = await api.get<any>('/analytics/overview', { params: { fromDate: startDate, toDate: endDate } });
        const raw = response.data || {};
        
        const total = raw.totalQuestions ?? 0;
        const answered = raw.answeredCount ?? raw.answeredQuestions ?? 0;
        const escalated = raw.escalatedCount ?? raw.escalated ?? 0;
        const unanswered = raw.noAnswerCount ?? raw.unansweredQuestions ?? 0;
        const today = raw.todayQuestions ?? raw.questionsToday ?? 0;
        const rate = raw.answerRate ?? raw.successRate ?? (total > 0 ? (answered / total) : 0);
        const avgMs = raw.avgResponseTimeMs ?? 0;
        const avgTimeFormatted = avgMs > 0 ? `${(avgMs / 1000).toFixed(2)}s` : '0.4s';

        return {
            totalQuestions: total,
            questionsToday: today,
            todayQuestions: today,
            answeredQuestions: answered,
            answeredCount: answered,
            unansweredQuestions: unanswered,
            noAnswerCount: unanswered,
            escalated: escalated,
            escalatedCount: escalated,
            successRate: rate,
            answerRate: rate,
            avgResponseTime: avgTimeFormatted,
            avgResponseTimeMs: avgMs,
            avgSimilarityScore: raw.avgSimilarityScore ?? 0
        };
    },
    getQuestionAnalytics: async (startDate?: string, endDate?: string): Promise<QuestionAnalytics[]> => {
        const response = await api.get<any>('/analytics/questions', { params: { fromDate: startDate, toDate: endDate } });
        const data = response.data;
        const list = data?.questions || (Array.isArray(data) ? data : []);
        return list;
    },
    getKnowledgeAnalytics: async (page = 1, pageSize = 8, startDate?: string, endDate?: string): Promise<PaginatedResult<KnowledgeAnalytics>> => {
        const response = await api.get<any>('/analytics/knowledge', { params: { page, pageSize, fromDate: startDate, toDate: endDate } });
        const data = response.data;
        const list = Array.isArray(data) ? data : (data?.items || []);
        const totalCount = data?.totalCount ?? list.length;
        const totalPages = data?.totalPages ?? Math.max(1, Math.ceil(totalCount / pageSize));
        const items = list.map((item: any) => ({
            id: item.scenarioId || item.categoryId || item.id,
            scenarioId: item.scenarioId,
            scenarioName: item.scenarioName || 'سيناريو عام',
            categoryName: item.categoryName || 'غير مصنف',
            usageCount: item.retrievalCount ?? item.usageCount ?? 0,
            avgSimilarityScore: item.avgSimilarityScore ?? 0
        }));
        return { items, totalCount, page, pageSize, totalPages };
    },
    getCategoryAnalytics: async (page = 1, pageSize = 5, startDate?: string, endDate?: string): Promise<PaginatedResult<CategoryAnalytics>> => {
        const response = await api.get<any>('/analytics/categories', { params: { page, pageSize, fromDate: startDate, toDate: endDate } });
        const data = response.data;
        const list = Array.isArray(data) ? data : (data?.items || []);
        const totalCount = data?.totalCount ?? list.length;
        const totalPages = data?.totalPages ?? Math.max(1, Math.ceil(totalCount / pageSize));
        return { items: list, totalCount, page, pageSize, totalPages };
    },
    getUnanswered: async (page = 1, pageSize = 8, sortOrder = 'desc', startDate?: string, endDate?: string): Promise<PaginatedResult<UnansweredQuestion>> => {
        const response = await api.get<any>('/analytics/unanswered', { params: { page, pageSize, sortOrder, fromDate: startDate, toDate: endDate } });
        const data = response.data;
        const list = data?.questions || (Array.isArray(data) ? data : (data?.items || []));
        const totalCount = data?.totalCount ?? list.length;
        const totalPages = data?.totalPages ?? Math.max(1, Math.ceil(totalCount / pageSize));
        const items = list.map((u: any) => ({
            id: u.id,
            question: u.questionText || u.question || '',
            questionText: u.questionText || u.question || '',
            timestamp: u.lastAsked || u.firstAsked || u.createdAt || new Date().toISOString(),
            frequency: u.frequency || 1,
            employeeId: u.employeeId,
            employeeName: u.employeeName,
            employeeEmail: u.employeeEmail
        }));
        return { items, totalCount, page, pageSize, totalPages };
    }
};

export default analyticsService;
